### ISSUE TITLE
More control over step traversal

### ISSUE THREAD

**Issue by kyse** [stepstone-tech/android-material-stepper#42]
Would be nice to see the ability to make the next/back/completed buttons Visibility.gone, and expose methods on the StepperLayout class to allow me to programmatically advance to the next step.  This would allow for things like where a step requires a user select a button out of a choice of buttons for example, and the code does something on button click, and once done, advance the step.  No need to have the user also click next.  Also, add the ability to prevent the tabs from updating the steps along the same lines.

While I can just grab the buttons and tabs container and set the button onClickListener to null, and overload the tabContainers listener to achieve the same result, I'd have to make the buttons text the same color as the nav bar, and use some styling to get the chevrons to be the same color effectively making them disappear.  However the bar is still there on the ui, which sucks if it's unneeded to begin with.

I can't just hide the buttons unfortunately since you update the visibility inside onUpdate, so would be nice to add a && shouldBeVisibile style able or something along those lines inside onUpdate, so I can just use styles in a theme to change the behavior base on individual stepperlayouts in my views.

I'm guessing in the mean time, I can grab the listeners before I set them to null so I can have access to the StepperLayouts private methods for actually advancing the steps?

**Comment by zawadz88** [comment]
Hi @kyse,
Correct me if I'm wrong, but does it mean that you would like to hide the navigation bars altogether?
If so, wouldn't it be actually easier for you to use a plain old ViewPager instead of this library?

You can also navigate to a selected step programmatically by calling StepperLayout#_setCurrentStepPosition(..)_

**Comment by kyse** [comment]
I was trying to make use of the graphics you have for the header, but perhaps your right, will look at the view pager.

**Comment by zawadz88** [comment]
Hi @kyse ,
Any luck? I was thinking about this and IMHO the library is all about the UI and not just about the step verification mechanism therefore.
If you want to hide them though you might get `ms_stepTabsContainer` and `ms_stepNavigation` by their IDs and set visibility to `GONE`. Going to the next step with step verification could be done by calling `StepperLayout#onTabClicked(position)`

**Comment by kyse** [comment]
Well, I went down the rabbit hole of creating a custom view for implementing a vertical stepper.  I like the look of it, and it works well for my intended purpose (couple buttons on one step, couple input text fields on the next step and done).  Found a form that's already created that has a ton of nested linearlayout and relativelaouts, and injects a confirmation step at the end, more lack of customization built in (sigh).

Went down the rabbit hole of trying to create a stepper myself to mimic the material design imp of a mobile vertical stepper.  Successfully created a view that: draws the oval shape or error drawable, draws the number, check or edit drawable on top of the circle using Paint and TextPaint.  Then for the implementation of the wrapper, I'm debating two different routes.  Expanding the StepCircle to additionally draw the title and subtitle text, declare it as a viewpager decorator, try and get a second decorator for the left line divider (if i can get horizontal gravity decorators to work), and a bottom decorator for the next steps.  Then use a combination of updating the states of the decorators and a transition control for the viewpager content for the fragment to represent closing one step and moving to the next step.

Either that, or just use a ConstraintLayout to wrap a frameview with the graphics/titles, and directly manage the fragments into the frameviews of each step by hosting the fragmentmanager in the adapter or a wrapper view to host/build each individual step view.

I feel like the first option would be more of what google intends, less hierarchy, controlling performance via controlling rendering of the drawables, and reuse existing functionality with viewpager, and viewpager's decorators.  Just not sure I can achieve the proper feel I desire since viewpager is really only meant for horizontal paging, and I believe the decorators are only meant to use the gravity vertical axes (TOP, BOTTOM).

But, it's a rabbit hole because I'm still learning (which is the point of side projects right?).

But I appreciate you response, I agree, these libraries are meant to supply developers with a means to a desired end, not restrict them in what they can do right?
